homeassistant/components/vacuum/demo.py

```diff
@@ -204,3 +212,118 @@ def send_command(self, command, params=None, **kwargs):
         self._status = 'Executing {}({})'.format(command, params)
         self._state = True
         self.schedule_update_ha_state()
+
+
+class StateDemoVacuum(StateVacuumDevice):
+    """Representation of a demo vacuum supporting states."""
+
+    def __init__(self, name):
+        """Initialize the vacuum."""
+        self._name = name
+        self._supported_features = SUPPORT_STATE_SERVICES
+        self._state = STATE_DOCKED
+        self._fan_speed = FAN_SPEEDS[1]
+        self._cleaned_area = 0
+        self._battery_level = 100
+
+    @property
+    def name(self):
+        """Return the name of the vacuum."""
+        return self._name
+
+    @property
+    def should_poll(self):
+        """No polling needed for a demo vacuum."""
+        return False
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._supported_features
+
+    @property
+    def state(self):
+        """Return the current state of the vacuum."""
+        return self._state
+
+    @property
+    def battery_level(self):
+        """Return the current battery level of the vacuum."""
+        if self.supported_features & SUPPORT_BATTERY == 0:
+            return
+
+        return max(0, min(100, self._battery_level))
+
+    @property
+    def fan_speed(self):
+        """Return the current fan speed of the vacuum."""
+        if self.supported_features & SUPPORT_FAN_SPEED == 0:
+            return
+
+        return self._fan_speed
+
+    @property
+    def fan_speed_list(self):
+        """Return the list of supported fan speeds."""
+        if self.supported_features & SUPPORT_FAN_SPEED == 0:
+            return
+        return FAN_SPEEDS
+
+    @property
+    def device_state_attributes(self):
+        """Return device state attributes."""
+        return {ATTR_CLEANED_AREA: round(self._cleaned_area, 2)}
+
+    def start_pause(self, **kwargs):
+        """Start, pause or resume the cleaning task."""
+        if self.supported_features & SUPPORT_PAUSE == 0:
```

Comment:
Doesn't this mean the vacuum cannot be started if it doesn't support pause?  The demo vacuum doesn't support pause so it should definitely be broken.

Suggest doing this below inside of if self._state == STATE_CLEANING: ...

However, this means that calling this method on a CLEANING vacuum which doesn't support pause will just do nothing.  Is that OK?

Reply:
True, I wanted to do a refactor of start_pause for all vacuums (splitting them up) in a separate PR, seeing as this currently isn't a problem (all current platforms support pause), this could be added in another platform in just a couple of lines of code.

Reply:
Oh, SUPPORT_STATE_SERVICES contains pause.  I'm way too tired.